apps/site/assets/ts/schedule/components/schedule-finder/UpcomingDepartures.tsx

```diff
@@ -1,30 +1,138 @@
-import React, { ReactElement } from "react";
+import React, { ReactElement, useEffect, useReducer } from "react";
+import { isNull } from "util";
 import {
   timeForCommuterRail,
   trackForCommuterRail,
   statusForCommuterRail
 } from "../../../helpers/prediction-helpers";
 import { modeIcon } from "../../../helpers/icon";
 import { modeBgClass } from "../../../stop/components/RoutePillList";
-import { Route } from "../../../__v3api";
-import { StopPrediction } from "../__schedule";
+import { Route, StopPrediction } from "../../../__v3api";
+import {
+  ScheduleWithFare,
+  ServiceScheduleByTrip,
+  ServiceScheduleInfo
+} from "../__schedule";
 import { breakTextAtSlash } from "../../../helpers/text";
+import { Accordion } from "../../components/schedule-finder/TableRow";
+import { SelectedDirection, SelectedOrigin } from "../ScheduleFinder";
+import { reducer } from "../../../helpers/fetch";
+
+type fetchPredictionsAction =
+  | { type: "FETCH_COMPLETE"; payload: StopPrediction[] }
+  | { type: "FETCH_ERROR" }
+  | { type: "FETCH_STARTED" };
 
-interface State {
+interface PredictionState {
   data: StopPrediction[] | null;
+  initial: StopPrediction[] | null;
   isLoading: boolean;
   error: boolean;
 }
 
 interface Props {
-  state: State;
+  tripData: ServiceScheduleInfo | null;
+  directionId: SelectedDirection;
+  stopId: string;
+  routeId: string;
+}
+
+interface UpcomingBusProps {
+  tripData: ServiceScheduleInfo;
+  routeId: string;
+  directionId: SelectedDirection;
+  stopId: string;
+}
+
+interface UpcomingCrProps {
+  tripData: ServiceScheduleInfo;
 }
 
-const hasPredictions = (stopPredictions: StopPrediction[]): boolean =>
+const reduceTrips = ({
+  trip_order: tripOrder,
+  by_trip: byTrip
+}: ServiceScheduleInfo): ServiceScheduleInfo => {
+  const tripIdsWithPredictions: string[] = [];
+  const tripsWithPredictions = tripOrder.reduce(
+    (obj: ServiceScheduleByTrip, tripId: string) => {
+      const trip = byTrip[tripId];
+      if (
+        trip.schedules.some(
+          (schedule, idx) => !isNull(schedule.prediction.prediction) && idx < 2
+        )
+      ) {
+        tripIdsWithPredictions.push(tripId);
+        // eslint-disable-next-line no-param-reassign
+        obj[tripId] = trip;
```

Comment:
I think so. @phildarnowsky what do you think.

Reply:
@digitalcora @amaisano I like immutable data a lot too. In this case, I assured myself that it was safe to do it the mutable way, "safe" in terms of not mutating that which ought not be mutated.

The reason is that the `obj` we're mutating starts out assigned to the empty object down on line 70, and we build it up entry by entry. The mutation is all confined to this object that only exists in this scope.

I did it this way because I didn't know the syntax for specifying a field name dynamically in a context like this. If what @digitalcora proposes works, let's run with that.